File: src/profile-logic/import/jfr-wasm/jafar/src/main/java/me/bechberger/jafar/web/converter/Categories.java
package me.bechberger.jafar.web.converter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Port of categories.ts (originally CategoryE.kt). Defines the fixed list of
 * category buckets used by the Firefox Profiler view, with mutable subcategory
 * lists that grow as new (category, subcategory) pairs are observed during
 * conversion.
 */
public final class Categories {

    public static final class Entry {
        public final String displayName;
        public final String color;
        public final List<String> subcategories;
        public final int index;

        Entry(String displayName, String color, List<String> subcategories, int index) {
            this.displayName = displayName;
            this.color = color;
            this.subcategories = subcategories;
            this.index = index;
        }
    }

    private static final List<Entry> ALL = new ArrayList<>();
    private static final Map<String, Entry> BY_NAME = new HashMap<>();

    public static final Entry OTHER;
    public static final Entry JAVA;
    public static final Entry NON_PROJECT_JAVA;
    public static final Entry GC;
    public static final Entry CPP;
    public static final Entry JFR;
    public static final Entry JAVA_APPLICATION;
    public static final Entry JAVA_APPLICATION_STATS;
    public static final Entry JVM_CLASSLOADING;
    public static final Entry JVM_CODE_CACHE;
    public static final Entry JVM_COMPILATION_OPT;
    public static final Entry JVM_COMPILATION;
    public static final Entry JVM_DIAGNOSTICS;
    public static final Entry JVM_FLAG;
    public static final Entry JVM_GC_COLLECTOR;
    public static final Entry JVM_GC_CONF;
    public static final Entry JVM_GC_DETAILED;
    public static final Entry JVM_GC_HEAP;
    public static final Entry JVM_GC_METASPACE;
    public static final Entry JVM_GC_PHASES;
    public static final Entry JVM_GC_REFERENCE;
    public static final Entry JVM_INTERNAL;
    public static final Entry JVM_PROFILING;
    public static final Entry JVM_RUNTIME_MODULES;
    public static final Entry JVM_RUNTIME_SAFEPOINT;
    public static final Entry JVM_RUNTIME_TABLES;
    public static final Entry JVM_RUNTIME;
    public static final Entry JVM;
    public static final Entry OS_MEMORY;
    public static final Entry OS_NETWORK;
    public static final Entry OS_PROCESS;
    public static final Entry OS;
    public static final Entry MISC;

    static {
        OTHER = create("Other", "grey", "Profiling", "Waiting");
        JAVA = create("Java", "blue", "Other", "Interpreted", "Compiled", "Native", "Inlined");
        NON_PROJECT_JAVA = create("Java (non-project)", "darkgray",
            "Other", "Interpreted", "Compiled", "Native", "Inlined");
        GC = create("GC", "orange", "Other");
        CPP = create("Native", "red", "Other");
        JFR = create("Flight Recorder", "lightgrey");
        JAVA_APPLICATION = create("Java Application", "red");
        JAVA_APPLICATION_STATS = create("Java Application, Statistics", "grey");
        JVM_CLASSLOADING = create("Java Virtual Machine, Class Loading", "brown");
        JVM_CODE_CACHE = create("Java Virtual Machine, Code Cache", "lightbrown");
        JVM_COMPILATION_OPT = create("Java Virtual Machine, Compiler, Optimization", "lightblue");
        JVM_COMPILATION = create("Java Virtual Machine, Compiler", "lightblue");
        JVM_DIAGNOSTICS = create("Java Virtual Machine, Diagnostics", "lightgrey");
        JVM_FLAG = create("Java Virtual Machine, Flag", "lightgrey");
        JVM_GC_COLLECTOR = create("Java Virtual Machine, GC, Collector", "orange");
        JVM_GC_CONF = create("Java Virtual Machine, GC, Configuration", "lightgrey");
        JVM_GC_DETAILED = create("Java Virtual Machine, GC, Detailed", "lightorange");
        JVM_GC_HEAP = create("Java Virtual Machine, GC, Heap", "lightorange");
        JVM_GC_METASPACE = create("Java Virtual Machine, GC, Metaspace", "lightorange");
        JVM_GC_PHASES = create("Java Virtual Machine, GC, Phases", "lightorange");
        JVM_GC_REFERENCE = create("Java Virtual Machine, GC, Reference", "lightorange");
        JVM_INTERNAL = create("Java Virtual Machine, Internal", "lightgrey");
        JVM_PROFILING = create("Java Virtual Machine, Profiling", "lightgrey");
        JVM_RUNTIME_MODULES = create("Java Virtual Machine, Runtime, Modules", "lightgrey");
        JVM_RUNTIME_SAFEPOINT = create("Java Virtual Machine, Runtime, Safepoint", "yellow");
        JVM_RUNTIME_TABLES = create("Java Virtual Machine, Runtime, Tables", "lightgrey");
        JVM_RUNTIME = create("Java Virtual Machine, Runtime", "green");
        JVM = create("Java Virtual Machine", "lightgrey");
        OS_MEMORY = create("Operating System, Memory", "lightgrey");
        OS_NETWORK = create("Operating System, Network", "lightgrey");
        OS_PROCESS = create("Operating System, Processor", "lightgrey");
        OS = create("Operating System", "lightgrey");
        MISC = create("Misc", "lightgrey", "Other");
    }

    private static Entry create(String displayName, String color, String... subcategories) {
        List<String> subs = new ArrayList<>();
        for (String s : subcategories) subs.add(s);
        Entry e = new Entry(displayName, color, subs, ALL.size());
        ALL.add(e);
        BY_NAME.put(displayName, e);
        return e;
    }

    /** Packed (categoryIndex, subcategoryIndex) — avoid int[2] allocation. Use
     *  {@link #subCat(long)} / {@link #subSub(long)} to unpack. Adds the
     *  subcategory if new. */
    public static long subPacked(Entry category, String subcategoryName) {
        int idx = category.subcategories.indexOf(subcategoryName);
        if (idx == -1) {
            category.subcategories.add(subcategoryName);
            idx = category.subcategories.size() - 1;
        }
        return ((long) category.index << 32) | (idx & 0xFFFFFFFFL);
    }

    public static int subCat(long packed) { return (int) (packed >>> 32); }
    public static int subSub(long packed) { return (int) packed; }

    /** Legacy wrapper kept for the cold paths that don't run per-frame. */
    public static int[] sub(Entry category, String subcategoryName) {
        long p = subPacked(category, subcategoryName);
        return new int[] { subCat(p), subSub(p) };
    }

    public static Entry fromCategoryName(String displayName) {
        Entry e = BY_NAME.get(displayName);
        return e != null ? e : OTHER;
    }

    public static List<Entry> all() {
        return ALL;
    }

    /** Append the categories array to the JSON output. */
    public static void writeCategoryList(JsonWriter w) {
        w.beginArray();
        for (Entry e : ALL) {
            w.beginObject();
            w.keyString("name", e.displayName);
            w.keyString("color", e.color);
            w.key("subcategories").beginArray();
            for (String s : e.subcategories) w.value(s);
            w.endArray();
            w.endObject();
        }
        w.endArray();
    }
}

File: src/profile-logic/import/jfr-converter/categories.ts
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// Port of CategoryE.kt

export interface Category {
  name: string;
  color: string;
  subcategories: string[];
}

interface CategoryEntry {
  displayName: string;
  color: string;
  subcategories: string[];
  index: number;
}

function cat(
  displayName: string,
  color: string,
  subcategories: string[] = []
): CategoryEntry {
  return { displayName, color, subcategories, index: 0 };
}

// Keep insertion order — index is assigned below.
const _CATEGORIES: CategoryEntry[] = [
  cat('Other', 'grey', ['Profiling', 'Waiting']),
  cat('Java', 'blue', ['Other', 'Interpreted', 'Compiled', 'Native', 'Inlined']),
  cat('Java (non-project)', 'darkgray', [
    'Other',
    'Interpreted',
    'Compiled',
    'Native',
    'Inlined',
  ]),
  cat('GC', 'orange', ['Other']),
  cat('Native', 'red', ['Other']),
  cat('Flight Recorder', 'lightgrey'),
  cat('Java Application', 'red'),
  cat('Java Application, Statistics', 'grey'),
  cat('Java Virtual Machine, Class Loading', 'brown'),
  cat('Java Virtual Machine, Code Cache', 'lightbrown'),
  cat('Java Virtual Machine, Compiler, Optimization', 'lightblue'),
  cat('Java Virtual Machine, Compiler', 'lightblue'),
  cat('Java Virtual Machine, Diagnostics', 'lightgrey'),
  cat('Java Virtual Machine, Flag', 'lightgrey'),
  cat('Java Virtual Machine, GC, Collector', 'orange'),
  cat('Java Virtual Machine, GC, Configuration', 'lightgrey'),
  cat('Java Virtual Machine, GC, Detailed', 'lightorange'),
  cat('Java Virtual Machine, GC, Heap', 'lightorange'),
  cat('Java Virtual Machine, GC, Metaspace', 'lightorange'),
  cat('Java Virtual Machine, GC, Phases', 'lightorange'),
  cat('Java Virtual Machine, GC, Reference', 'lightorange'),
  cat('Java Virtual Machine, Internal', 'lightgrey'),
  cat('Java Virtual Machine, Profiling', 'lightgrey'),
  cat('Java Virtual Machine, Runtime, Modules', 'lightgrey'),
  cat('Java Virtual Machine, Runtime, Safepoint', 'yellow'),
  cat('Java Virtual Machine, Runtime, Tables', 'lightgrey'),
  cat('Java Virtual Machine, Runtime', 'green'),
  cat('Java Virtual Machine', 'lightgrey'),
  cat('Operating System, Memory', 'lightgrey'),
  cat('Operating System, Network', 'lightgrey'),
  cat('Operating System, Processor', 'lightgrey'),
  cat('Operating System', 'lightgrey'),
  cat('Misc', 'lightgrey', ['Other']),
];

_CATEGORIES.forEach((c, i) => (c.index = i));

const _CATEGORY_MAP = new Map<string, CategoryEntry>(
  _CATEGORIES.map((c) => [c.displayName, c])
);

// Named accessors matching the Kotlin enum members used in Processor/Tables
export const CategoryE = {
  OTHER: _CATEGORIES[0],
  JAVA: _CATEGORIES[1],
  NON_PROJECT_JAVA: _CATEGORIES[2],
  GC: _CATEGORIES[3],
  CPP: _CATEGORIES[4],
  JFR: _CATEGORIES[5],
  JAVA_APPLICATION: _CATEGORIES[6],
  JAVA_APPLICATION_STATS: _CATEGORIES[7],
  JVM_CLASSLOADING: _CATEGORIES[8],
  JVM_CODE_CACHE: _CATEGORIES[9],
  JVM_COMPILATION_OPT: _CATEGORIES[10],
  JVM_COMPILATION: _CATEGORIES[11],
  JVM_DIAGNOSTICS: _CATEGORIES[12],
  JVM_FLAG: _CATEGORIES[13],
  JVM_GC_COLLECTOR: _CATEGORIES[14],
  JVM_GC_CONF: _CATEGORIES[15],
  JVM_GC_DETAILED: _CATEGORIES[16],
  JVM_GC_HEAP: _CATEGORIES[17],
  JVM_GC_METASPACE: _CATEGORIES[18],
  JVM_GC_PHASES: _CATEGORIES[19],
  JVM_GC_REFERENCE: _CATEGORIES[20],
  JVM_INTERNAL: _CATEGORIES[21],
  JVM_PROFILING: _CATEGORIES[22],
  JVM_RUNTIME_MODULES: _CATEGORIES[23],
  JVM_RUNTIME_SAFEPOINT: _CATEGORIES[24],
  JVM_RUNTIME_TABLES: _CATEGORIES[25],
  JVM_RUNTIME: _CATEGORIES[26],
  JVM: _CATEGORIES[27],
  OS_MEMORY: _CATEGORIES[28],
  OS_NETWORK: _CATEGORIES[29],
  OS_PROCESS: _CATEGORIES[30],
  OS: _CATEGORIES[31],
  MISC: _CATEGORIES[32],
} as const;

// Returns (categoryIndex, subcategoryIndex), adding subcategory if new
export function sub(
  category: CategoryEntry,
  subcategoryName: string
): [number, number] {
  let idx = category.subcategories.indexOf(subcategoryName);
  if (idx === -1) {
    category.subcategories.push(subcategoryName);
    idx = category.subcategories.length - 1;
  }
  return [category.index, idx];
}

export function fromCategoryName(displayName: string): CategoryEntry {
  return _CATEGORY_MAP.get(displayName) ?? CategoryE.OTHER;
}

export function toCategoryList(): Category[] {
  return _CATEGORIES.map((c) => ({
    name: c.displayName,
    color: c.color,
    subcategories: [...c.subcategories],
  }));
}

export const MISC_OTHER = sub(CategoryE.MISC, 'Other');
